solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -58,6 +62,67 @@ def apply_validation(obs_df, qfilter, handle_func):
         return validated_obs, counts
 
 
+def apply_fill(fx_data, missing_forecast, start, end, interval_length):
+    """
+    Apply fill procedure to the data from the start to end timestamps.
+
+    Parameters
+    ----------
+    fx_data : pandas.Series or pandas.DataFrame
+        Forecast data with pandas.DatetimeIndex.
+    missing_forecast : str
+        Indicates what process to use for handling missing forecasts.
+        Currently supports : 'drop', 'forward', and bool or numeric value.
+    start : pandas.Timestamp
+    end : pandas.Timestamp
+    interval_length : pandas.Timedelta
+
+    Returns
+    -------
+    pandas.Series or pandas.DataFrame
+    int : Number of values filled or dropped.
+
+    TODO
+    ----
+    - Add support for 'clearsky' fillin.
+
+    """
+    try:
+        orig_dtype = fx_data.dtype
+    except AttributeError:
+        # DataFrame must be all same dtypes
+        orig_dtype = set(fx_data.dtypes).pop()
+    missing_forecast = str(missing_forecast)
+    # Create full datetime range at resolution
+    full_dt_index = pd.date_range(start=start, end=end, freq=interval_length,
+                                  name=fx_data.index.name)
+
+    if missing_forecast == 'drop':
+        if isinstance(fx_data, pd.DataFrame):
+            count = fx_data.isna().any(axis=1).sum() * fx_data.shape[1]
+        else:
+            count = fx_data.isna().sum()
+        fx_data_proc = fx_data.dropna(how='any')
+    elif missing_forecast == 'forward':
+        fx_data_proc = fx_data.reindex(index=full_dt_index)
+        count = fx_data_proc.isna().sum()
+        fx_data_proc.fillna(method='ffill', inplace=True)
+        fx_data_proc.fillna(value=0, inplace=True)  # leading gap
+    elif missing_forecast.lstrip('-').replace('.', '').isnumeric():
+        fill_value = pd.to_numeric(missing_forecast).astype(orig_dtype)
+        fx_data_proc = fx_data.reindex(index=full_dt_index)
+        count = fx_data_proc.isna().sum()
+        fx_data_proc.fillna(value=fill_value, inplace=True)
+    else:
+        raise ValueError(f"Unsupported forecast missing data procedure: "
+                         f"{missing_forecast}")
+
+    if isinstance(count, pd.Series):
+        count = count.sum()
```

Comment:
Added a comment. If using `sum` on a `DataFrame` it will return a `Series` of sums and need to sum overall of those to get the total for the whole `DataFrame`.

Reply:
Also added a comment for `drop` that any row that has missing data is dropped for all columns.  Let me know it this kind of stuff should be documented in the metrics or Use Case?